src/libgrate/matrix.c

```diff
@@ -143,7 +143,7 @@ void mat4_perspective(struct mat4 *m, float fov, float aspect,
 	m->xx = cotangent / aspect;
 	m->yy = cotangent;
 	m->zz = -(far + near) / depth;
-	m->zw = -1.0f;
-	m->wz = -2.0f * near * far / depth;
+	m->zw = -2.0f * near * far / depth;
+	m->wz = -1.0f;
```

Comment:
Row-major vs column-major only matters when you write matrices mathematically, it doesn't dictate the memory-layout of the matrix. It's a notation-thing, not a memory-layout thing.

An GLSL supports both column-vectors *and* row-vectors, you can do `mat4 * vec4` or `vec4 * mat4`, and they do different operations.